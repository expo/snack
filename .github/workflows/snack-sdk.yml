name: SnackSDK

defaults:
  run:
    working-directory: packages/snack-sdk

on:
  workflow_dispatch: {}
  push:
    branches: [main]
    paths:
      - .github/workflows/snack-sdk.yml
      - packages/snack-sdk/**
      - .eslint*
      - .prettier*
      - yarn.lock
  pull_request:
    branches: [main]
    paths:
      - .github/workflows/snack-sdk.yml
      - packages/snack-sdk/**
      - .eslint*
      - .prettier*
      - yarn.lock

jobs:
  build:
    runs-on: ubuntu-18.04
    steps:
      - uses: actions/checkout@v2
      - name: Get yarn cache directory path
        id: yarn-cache-dir-path
        run: echo "::set-output name=dir::$(yarn cache dir)"
      - uses: actions/cache@v1
        with:
          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
          key: ${{ runner.os }}-yarn-${{ hashFiles('yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-
      - run: yarn install --ignore-scripts --frozen-lockfile
      - run: yarn test --maxWorkers 1
      - run: yarn lint --max-warnings 0
      - run: yarn build
      - run: yarn doc:ci
