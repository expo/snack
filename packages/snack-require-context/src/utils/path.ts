/**
 * Sanitize a file path from Babel or Snack Runtime.
 * This removes any leading `/`, `./`, or `module://` prefixes.
 */
export function sanitizeFilePath(filePath: string) {
  // Remove starting `/`, `./`, or `module://` prefixes
  return filePath.replace(/^(module:\/{1,2}|\.?\/)/, '');
}
