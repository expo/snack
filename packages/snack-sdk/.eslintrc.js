const { jsExtensions, tsExtensions } = require('eslint-config-universe/shared/extensions');

module.exports = {
  extends: 'universe/node',
  // This project does not contain babel, force all files to use the typescript parser instead.
  parser: '@typescript-eslint/parser',
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': [...jsExtensions, ...tsExtensions],
    },
  },
};
