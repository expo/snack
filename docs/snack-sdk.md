# Snack SDK Documentation <!-- omit in toc -->

The Expo Snack SDK. Use this to create a custom web interface for https://snack.expo.dev/. 


## Index <!-- omit in toc -->

- [Basic usage](#basic-usage)
- [Files, dependencies and the state](#files-dependencies-and-the-state)
  - [Updating files](#updating-files)
  - [Using assets](#using-assets)
  - [Using dependencies](#using-dependencies)
  - [Changing the SDK version](#changing-the-sdk-version)
  - [Getting the state](#getting-the-state)
  - [Missing dependencies](#missing-dependencies)
- [Going online, debounced updates and connected clients](#going-online-debounced-updates-and-connected-clients)
  - [Going online](#going-online)
  - [Sending code changes manually or debounced](#sending-code-changes-manually-or-debounced)
  - [DeviceId's and connected accounts](#deviceids-and-connected-accounts)
  - [Connected clients](#connected-clients)
  - [Requesting previews](#requesting-previews)
- [Saving the Snack](#saving-the-snack)
  - [Downloading as a Zip file](#downloading-as-a-zip-file)
- [Transports](#transports)
  - [Previewing the Snack on Web](#previewing-the-snack-on-web)
- [Example App](#example-app)
- [API Reference](#api-reference)

# Basic usage

```sh
yarn add snack-sdk
```

```ts
import { Snack } from 'snack-sdk';

// Create Snack
const snack = new Snack({
  files: {
    'App.js': {
      type: 'CODE',
      contents: `
import * as React from 'react';
import { View, Text } from 'react-native';

export default () => (
  <View style={{flex: 1, justifyContent: 'center'}}>
    <Text style={{fontSize: 20, textAlign: 'center'}}>
      Hello Snack!
    </Text>
  </View>
);
`
    }
  }
});

// Make the Snack available online
snack.setOnline(true);
const { url } = await snack.getStateAsync();

// You can now use the url and show it as a QR code
// to open the Snack in the Expo client.

// Stop Snack when done
snack.setOnline(false);
```

# Files, dependencies and the state

A Snack consists of a collection of files and dependencies which make up the code of the application. You can choose to use either JavaScript or TypeScript by using one of the following filenames as the entry-point for the Snack.

- `App.js` (JavaScript)
- `App.tsx` (TypeScript)

## Updating files

When the code is updated, any connected clients will be automatically updated as well.

To add or update a file, use:

```ts
snack.updateFiles({
  'App.js': {
    type: 'CODE',
    contents: `console.log("Hello Snack");`,
  },
  'README.md': {
    type: 'CODE',
    contents: `# MyAwesomeSnack`,
  },
});
```

To delete a file, specify `null` as the value:

```ts
snack.updateFiles({
  'README.md': null,
});
```

## Using assets

Assets are resources that are available through a url, for instance images or fonts. Just as code-files, assets can be set using the constructor or the `updateFiles` method.

```ts
snack.updateFiles({
  'assets/logo.png': {
    type: 'ASSET',
    contents: 'https://mysite/logo.png',
  },
});
```

Assets can also be added as a [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) or [File](https://developer.mozilla.org/en-US/docs/Web/API/File) object, after which they are automatically uploaded and converted into a URL.

```ts
const blob = new Blob(...);

snack.updateFiles({
  'assets/logo.png': {
    type: 'ASSET',
    contents: blob
  }
});

// Wait for the upload to complete
const { files } = await snack.getStateAsync();
console.log(files['assets/logo.png'].contents); // string -> "https://..."
```

## Using dependencies

When using Snack, various common dependencies are already included such as `react-native` and `expo`. Other dependencies need to be added before those packages can be used. It is advised to add all the dependencies that are used in the code. Snack will automatically ignore the dependency if it is already preloaded into the snack-runtime.

```ts
import { Snack } from 'snack-sdk';

// Create Snack
const snack = new Snack({
  dependencies: {
    'expo-linear-gradient': {
      version: '8.2.1'
    }
  },
  files: {
    'App.js': {
      type: 'CODE',
      contents: `
import * as React from 'react';
import { LinearGradient } from 'expo-linear-gradient';

export default () => (
  <LinearGradient style={{flex: 1}} colors={['red', 'white', 'blue']} />
);
`
    }
  }
});
```

Similar to `updateFiles`, dependencies can be added or removed using the `updateDependencies` method.

```ts
snack.updateDependencies({
  'expo-haptics': { version: '8.2.1' }, // Add haptics
  'expo-linear-gradient': null // Remove linear gradient
});
```

When dependencies are added, Snack will resolve the semantic version on NPM and create a pre-bundled version of that package. In some cases this can take a while, for instance when a new version of a package has been released. You can wait for the resolving/bundling to complete using the `getStateAsync` method. Once a dependency has been resolved, you may store and re-use the `handle` so the package doesn't need to be resolved/bundled again.

```ts
const snack = new Snack({
  dependencies: {
    'expo-font': { version: '8.2.1' }
  }
});

// Wait for the dependencies to be resolved
const { dependencies } = await snack.getStateAsync();
console.log(`expo-font handle ${dependencies['expo-font'].handle}`); // string -> https://...

// Create a Snack with pre-resolved dependencies
const newSnack = new Snack({
  dependencies
});
```

## Changing the SDK version

A specific SDK version can be specified in the constructor or through the `setSDKVersion` method.

```ts
import { Snack } from 'snack-sdk';

const snack = new Snack({
  sdkVersion: '36.0.0' // Optional SDK version to use
});

// Upgrade to a newer sdk
snack.setSDKVersion('38.0.0');
```

> Note that when the SDK version is changed, the `url` is also changed and any connected clients need to reconnect to the new `url`.

## Getting the state

The current state of the Snack can be accessed through the `getState()` method. The state contains all the code, assets, dependencies, SDK version, etc; and can be used to access the contents of the Snack. The returned state is readonly and should **never** be changed externally.

```ts
const snack = new Snack({
  name: 'Wonderful orange',
  description: `It's a wonderful world`,
  sdkVersion: '37.0.0'
});

const { name, description, sdkVersion } = snack.getState();
console.log(name, description, sdkVersion);
```

When adding dependencies that need to be resolved, or asset files that need to be uploaded, you can wait for these kinds of operations to complete by using the `getStateAsync()` method. It waits for any pending operations to complete before returning the state.

```ts
// Add dependencies and assets for upload
snack.updateDependencies({
  'expo-contacts': { version: '9.0.0' }
});
snack.updateFiles({
  'assets/logo.png': {
    type: 'CODE',
    contents: new Blob(...)
  }
});

// Wait for the operations to complete
const { dependencies, files } = await snack.getStateAsync();
```

## Missing dependencies

When Snack detects that dependencies are missing, it reports these in the `missingDependencies` field of the state.
For instance, dependency `@react-navigation/stack` has peer-dependencies on `@react-navigation/native` and `react-native-screens`. When adding this dependency, `missingDependencies` will contain these missing peer dependencies as well as their "wanted" versions.

```ts
const snack = new Snack({
  '@react-navigation/stack': { version: '*' }
});
const state = await snack.getStateSsync();
console.log(state.missingDependencies);
/* {
  "@react-navigation/native": {
    dependents: ["@react-navigation/stack"],
    wantedVersion: 'x.x.x'
  },
  "react-native-screens": {
    dependents: ["@react-navigation/stack"],
    wantedVersion: 'x.x.x'
  },
  ...
} */
```

To fix missing dependencies, simply add them to the dependencies.

```ts
const { missingDependencies } = snack.getState();
const dependencies: SnackDependencies = {};
for (const name in snack.getState().missingDependencies) {
  dependencies[name] = {
    version: snack.getState().missingDependencies[name].wantedVersion
  };
}
snack.updateDependencies(dependencies);
```


# Going online, debounced updates and connected clients

## Going online

When a Snack is created, it is by default not online and clients are not able to connect. By setting the Snack to `online`, it advertises itself and Expo clients are able to connect. Use the `online` option in the constructor or the `setOnline` method to turn the online mode on or off.

```ts
const snack = new Snack({
  files: { ... }
  // online: true/false
});

snack.setOnline(true);

// console.log('URL' + snack.getState().url);
```

Once online, the `url` contains the unique address of the Snack and can be used to generate a QR code. This url stays the same for the lifetime of the Snack session, with exception of when the SDK version is changed. Changing the SDK version requires a different version of the snack-runtime to be loaded and will generate a different `url`.

## Sending code changes manually or debounced

By default, code changes are sent to the connected clients as quickly as possible. When editing code, this may however lead to a lot of communication and unwanted errors. When editing, it is therefore recommended to wait a short while (debounce) before sending the code changes; or to send them manually.

Set `codeChangesDelay` to a positive value to enable debounced code updates:

```ts
const snack = new Snack({
  online: true,
  files: { ... },
  codeChangesDelay: 1000 // milliseconds
});

snack.updateFiles({ ... });

// code changes are send after 1 second
```

Automatic code changes can also be turned off and triggered explicitely:

```ts
const snack = new Snack({
  online: true,
  files: { ... },
  codeChangesDelay: -1 // disable automatic code changes
});

// Updating files will not trigger any updates in the connected clients
snack.updateFiles({ ... });

// Send the code changes to the connected clients
snack.sendCodeChanges();
```

## DeviceId's and connected accounts

Instead of using the `url` and a QR-code, it is also possible to associate the Snack with a particular device-id or a user account. This will cause the Snack to automatically show up in the "Recently in Development" section of the Expo client.

To advertise the Snack on a specific device, use the `deviceId` field in the constructor or the `setDeviceId` method.

```ts
snack.setDeviceId('4321-1234');
```

Or set the account credentials to advertise it on all Expo clients using that account.

```ts
snack.setUser({ sessionSecret: '...' });
```

## Connected clients

Any clients that connect to the Snack will show up in the `connectedClients` field of the state.

```ts
snack.addStateListener((state, prevState) => {
  if (state.connectedClients !== prevState.connectedClients) {
    for (const key in state.connectedClients) {
      if (!prevState.connectedClients[key]) {
        console.log('A client has connected! ' + state.connectedClients[key].platform);
      }
    }
  }
});
```

## Requesting previews

When one or more clients are connected, the `getPreviewAsync` method can be used to request a visual preview of the Snack app.

```ts
const connectedClients = await snack.getPreviewAsync();

Object.values(connectedClients).forEach(client => {
  console.log(
    `Preview ${client.platform}, url: ${client.previewURL}, time: ${client.previewTimestamp}`
  );
});
```

# Saving the Snack

To save a Snack use the `saveAsync` method. It saves the Snack to the Expo servers and returns the Snack `id` and `url`.

```ts
const snack = new Snack({
  files: { ... },
  dependencies: { ... }
});

const { id, url } = await snack.saveAsync();
console.log(url); // "exp://exp.host/@jsghakdshgs"
```

To save the Snack to your user account, set the user in the constructor or using `setUser`.

```ts
const snack = new Snack({ user: { sessionSecret: '...' } });
snack.setUser({ sessionSecret: '...' });

// The snack will be automatically saved to the user account
await snack.saveAsync();

// To save an anonymously, specify the `ignoreUser` option
await snack.saveAsync({
  ignoreUser: true
});
```

To save a draft, use:

```ts
await snack.saveAsync({
  isDraft: true
});
```

## Downloading as a Zip file

Once a Snack has been saved, it can be downloaded as a Zip file. The `getDownloadURLAsync` method returns the URL through which the Snack can be downloaded. It also automatically save the Snack if it has any unsaved changes.

```ts
const url = snack.getDownloadURLAsync();
console.log('Download URL: ' + url); // https://exp.host/--/api/v2/snack/download/12345678
```

# Transports

The snack-sdk communicates with the Expo Runtime using Transports. When setting `online` to true, the default PubNub based transport is enabled and it is possible to connect to the Snack using the Expo Client. Additionally, a "web-preview" transport may be created, which communicates with the Snack web-player. The web-player is the Expo Runtime running on the web using [react-native-web](https://github.com/necolas/react-native-web).

## Previewing the Snack on Web

To use web-preview, create an iframe and pass it's `contentWindow` ref to the Snack. This creates a `webplayer` transport and `webPreviewURL` will be updated with the URL for the iframe. This URL may change, for instance when changing the SDK version, so make sure to update the iframe source.

> Web-preview is supported as of SDK 40. On older SDKs, `webPreviewURL` will always be `undefined`.

```jsx
import * as React from 'react';
import { Snack } from 'snack-sdk';

export default () => {
  const webPreviewRef = React.useRef(null);
  const [snack] = React.useState(() =>
    new Snack({
      ...
      webPreviewRef,
    })
  );
  const { webPreviewURL } = snack.getState();
  return (
    <div>
      ...
      <iframe
        ref={(c) => (webPreviewRef.current = c?.contentWindow ?? null)}
        src={webPreviewURL}
        allow="geolocation; camera; microphone"
      />
    </div>
  );
};
```

See the [Example App](/packages/snack-sdk/example) for a working demo.


# Example App

Head over to the [/packages/snack-sdk/example](../packages/snack-sdk/example) App for a live demo.

To run it, use:

```
yarn build
cd ./example
yarn
yarn start
```

# API Reference

[API Reference documentation](./snack-sdk-api/README.md)
