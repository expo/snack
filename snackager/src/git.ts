import { Request, Response } from 'express';
import GitUrlParse from 'git-url-parse';
import querystring from 'querystring';
import {
  isModulePreloaded,
  Snack,
  getSupportedSDKVersions,
  defaultConfig,
  SDKVersion,
} from 'snack-sdk';

import config from './config';
import logger from './logger';
import { getGitSnackObj } from './utils/convertRepoToSnack';

type Config = {
  repo: string;
  subpath?: string;
  branch?: string;
  hash?: string;
  noCache?: boolean;
};

export default async function git(req: Request, res: Response): Promise<void> {
  // @ts-ignore
  const parsed = querystring.parse(req._parsedUrl['query']);

  if (!parsed) {
    res.status(400);
    res.end('Failed to parse request');
    return;
  }

  try {
    logger.info(parsed, `importing repository ${parsed.repo}/${parsed.branch}/${parsed.subpath}`);
    const snackId = await importAsync({
      // @ts-ignore TODO: check if `parsed.repo` can contain `string[]` instead of `string`
      repo: parsed.repo,
      // @ts-ignore TODO: check if `parsed.subpath` can contain `string[]` instead of `string`
      subpath: parsed.subpath,
      // @ts-ignore TODO: check if `parsed.branch` can contain `string[]` instead of `string`
      branch: parsed.branch,
      // @ts-ignore TODO: check if `parsed.hash` can contain `string[]` instead of `string`
      hash: parsed.hash,
      // @ts-ignore TODO: check if `parsed.noCache` can contain `string[]` instead of `string`
      noCache: parsed.noCache && parsed.noCache !== '0',
    });
    logger.info(parsed, `import complete, available at Snack id ${snackId}`);
    res.status(200);
    res.end(snackId);
  } catch (e) {
    res.status(500);
    res.end('Failed to create snack: ' + e.message);
  }
}

export async function importAsync(options: Config): Promise<string> {
  let { repo, subpath, branch, hash, noCache } = options;
  if (subpath && subpath !== '/') {
    // remove leading and trailing forward slashes
    subpath = subpath.replace(/^\/+|\/+$/g, '');
  }

  const { files, dependencies, sdkVersion: uncheckedSDKVersion, date } = await getGitSnackObj(
    repo,
    subpath,
    branch,
    hash,
    noCache
  );

  // Verify SDK version
  let sdkVersion: SDKVersion = uncheckedSDKVersion as SDKVersion;
  if (!getSupportedSDKVersions().includes(sdkVersion)) {
    console.warn(
      `Unsupported SDK version "${sdkVersion}". Using "${defaultConfig.sdkVersion}" instead.`
    );
    sdkVersion = defaultConfig.sdkVersion;
  }

  // Format name
  // TODO: add support for commit hash specification
  const parsed = GitUrlParse(repo);
  let name =
    parsed.protocol === 'file' ? parsed.name : `${parsed.resource}/${parsed.owner}/${parsed.name}`;
  if (subpath && subpath !== '/') {
    name += `:${subpath}`;
  }
  if (branch && branch !== 'master' && branch !== 'main') {
    name += `@${branch}`;
  }

  // Format description
  // TODO: check if we can use date-fns for better date parsing/manipulation
  const dateParts = date.split(' ');
  const monthday = dateParts.slice(1, 3).join(' ');
  const year = dateParts[4];
  const description = `${name} @ ${monthday}, ${year}`;

  // Create a disabled Snack session. This Snack does not automatically
  // resolve dependencies, nor does it query the wanted package versions.
  const snack = new Snack({
    verbose: process.env.NODE_ENV === 'development',
    disabled: true,
    apiURL: config.api.url,
    snackagerURL: config.url,
    files,
    // Filter out all preloaded dependencies and format
    dependencies: Object.fromEntries(
      Object.entries(dependencies)
        .filter(([name]) => !isModulePreloaded(name, sdkVersion, true))
        .map(([name, version]) => [name, { version }])
    ),
    sdkVersion,
    name,
    description,
    user: { sessionSecret: process.env.GIT_SESSION_SECRET },
  });

  logger.info(options, 'saving repository');
  const { id } = await snack.saveAsync();
  return id;
}
