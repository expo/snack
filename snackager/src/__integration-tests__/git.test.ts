import spawnAsync from '@expo/spawn-async';
import FormData from 'form-data';
import fs from 'fs';
import fetch from 'node-fetch';
import os from 'os';
import path from 'path';
import { SnackFiles } from 'snack-sdk';
import util from 'util';

import { importAsync } from '../git';

// TODO: replace rimraf with fs.rm once node 14.40 lands
const rimraf = util.promisify(require('rimraf'));

const SAVE_ID = '7777771777';
const ASSET_URL = 'https://myuploadedasset.com/777';

jest.setTimeout(50000);
jest.mock('node-fetch', () => {
  return jest.fn(async (url: string, options: Request) => {
    let json = {};
    if (url.endsWith('bundledNativeModules.json')) {
      json = {};
    } else if (url.endsWith('/save')) {
      // @ts-ignore Body is rewritten as json so the call scan be compared to a snapshot
      options.body = JSON.parse(options.body);
      json = { id: SAVE_ID };
    } else if (url.endsWith('/uploadAsset')) {
      json = { url: ASSET_URL };
    }
    return {
      ok: true,
      status: 200,
      json: async () => json,
    };
  });
});
// Workaround for snack-sdk's ponyfill using their own node-fetch version
jest.mock('fetch-ponyfill', () => () => ({ fetch }));

// @ts-ignore Conversion of type 'typeof fetch' to type 'Mock<any, any>' may be a mistake because neither type sufficiently overlaps with the other
const mockedFetch = fetch as jest.Mock;

beforeEach(() => {
  mockedFetch.mockClear();
});

describe('git', () => {
  it('imports basic repository', async () => {
    const repoPath = await createRepo({
      name: 'test1',
      sdkVersion: '39.0.0',
    });
    const id = await importAsync({
      repo: repoPath,
      noCache: true,
    });
    expect(id).toBe(SAVE_ID);
    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
  });

  it('imports repository with no sdkVersion', async () => {
    const repoPath = await createRepo({
      name: 'test2',
      sdkVersion: undefined,
    });
    const id = await importAsync({
      repo: repoPath,
      noCache: true,
    });
    expect(id).toBe(SAVE_ID);
    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
  });

  it('imports repository with too old sdkVersion', async () => {
    const repoPath = await createRepo({
      name: 'test3',
      sdkVersion: '22.0.0',
    });
    const id = await importAsync({
      repo: repoPath,
      noCache: true,
    });
    expect(id).toBe(SAVE_ID);
    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
  });

  it('imports repository without app.json', async () => {
    const repoPath = await createRepo({
      name: 'some-example',
      sdkVersion: '40.0.0',
      appConfig: 'none',
    });
    const id = await importAsync({
      repo: repoPath,
      noCache: true,
    });
    expect(id).toBe(SAVE_ID);
    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
  });

  it('imports repository with app.config.js', async () => {
    const repoPath = await createRepo({
      name: 'some-example',
      sdkVersion: '39.0.0',
      appConfig: 'app.config.js',
    });
    const id = await importAsync({
      repo: repoPath,
      noCache: true,
    });
    expect(id).toBe(SAVE_ID);
    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
  });

  it('imports repository with assets', async () => {
    const repoPath = await createRepo({
      name: 'some-example',
      sdkVersion: '39.0.0',
      extraFiles: {
        'assets/image.png': {
          type: 'ASSET',
          contents: createAsset('image.png', 'hi'),
        },
      },
    });
    const id = await importAsync({
      repo: repoPath,
      noCache: true,
    });
    expect(id).toBe(SAVE_ID);
    expect(mockedFetch).toHaveBeenCalledTimes(2);
    expect(mockedFetch.mock.calls[0][0]).toMatchSnapshot();
    expect(mockedFetch.mock.calls[1]).toMatchSnapshot();
  });

  it('imports repository created with main branch', async () => {
    const repoPath = await createRepo({
      name: 'test1',
      sdkVersion: '39.0.0',
      branch: 'main',
    });
    const id = await importAsync({
      repo: repoPath,
      noCache: true,
    });
    expect(id).toBe(SAVE_ID);
    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
  });

  it('imports repository with custom branch', async () => {
    const repoPath = await createRepo({
      name: 'test1',
      sdkVersion: '39.0.0',
      branch: 'feature-a',
    });
    const id = await importAsync({
      repo: repoPath,
      branch: 'feature-a',
      noCache: true,
    });
    expect(id).toBe(SAVE_ID);
    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
  });

  it('fails to import repository with wrong branch', async () => {
    const repoPath = await createRepo({
      name: 'test1',
      sdkVersion: '39.0.0',
      branch: 'feature-a',
    });
    expect(
      importAsync({
        repo: repoPath,
        branch: 'feature-b',
        noCache: true,
      })
    ).rejects.toEqual(new Error(`Error cloning repo: git exited with non-zero code: 128`));
  });
});

async function createRepo(config: {
  name: string;
  sdkVersion: string | undefined;
  dependencies?: { [name: string]: string };
  appConfig?: 'app.json' | 'app.config.js' | 'none';
  extraFiles?: SnackFiles;
  branch?: string;
}): Promise<string> {
  const {
    name,
    sdkVersion,
    dependencies,
    appConfig = 'app.json',
    extraFiles = {},
    branch,
  } = config;
  const expoConfig = { name, sdkVersion };
  const files = {
    ...(appConfig === 'app.json' ? { 'app.json': JSON.stringify({ expo: expoConfig }) } : {}),
    ...(appConfig === 'app.config.js'
      ? { 'app.config.js': `export default ${JSON.stringify(expoConfig)};` }
      : {}),
    'App.js': `console.log('Hello World');`,
    ...extraFiles,
    'package.json': JSON.stringify({
      name,
      dependencies: dependencies ?? {
        'react-native': '0.62.0',
        expo: '38.0.10',
        'expo-asset': 'latest',
      },
    }),
  };

  // Create project dir with files
  const workdir = path.join(
    fs.realpathSync(os.tmpdir()),
    'snackager',
    '__integration-tests__',
    'git',
    name
  );
  try {
    await rimraf(workdir);
  } catch (e) {}
  fs.mkdirSync(workdir, { recursive: true });
  Object.keys(files).map((filename) => {
    const fullpath = path.join(workdir, filename);
    if (!fs.existsSync(path.dirname(fullpath))) {
      fs.mkdirSync(path.dirname(fullpath));
    }
    fs.writeFileSync(fullpath, files[filename]);
  });

  // Convert it into a git repository
  const commitDate = '2020-01-01T12:00:00.000Z';
  await spawnAsync('git', ['init'], { cwd: workdir });
  await spawnAsync('git', ['add', '--all'], { cwd: workdir });
  await spawnAsync('git', ['config', 'user.email', 'ci@expo.com'], { cwd: workdir });
  await spawnAsync('git', ['config', 'user.name', 'CI'], { cwd: workdir });
  await spawnAsync('git', ['add', '--all'], { cwd: workdir });
  await spawnAsync('git', ['commit', '--date', commitDate, '--message', 'Initial'], {
    cwd: workdir,
    env: {
      ...process.env,
      GIT_COMMITTER_DATE: commitDate,
    },
  });

  // Rename branch afterwards. This approach works for git versions prior to 2.28
  // https://stackoverflow.com/a/42871621
  if (branch) {
    await spawnAsync('git', ['branch', '-m', branch], { cwd: workdir });
  }

  return workdir;
}

function createAsset(name: string, contents: string): any {
  const formData = new FormData();
  formData.append('asset', contents, name);
  return formData;
}
