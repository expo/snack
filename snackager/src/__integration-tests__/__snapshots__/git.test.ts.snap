// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git imports basic repository 1`] = `
Array [
  "noop/--/api/v2/snack/save",
  Object {
    "body": Object {
      "code": Object {
        "App.js": Object {
          "contents": "console.log('Hello World');",
          "type": "CODE",
        },
        "app.json": Object {
          "contents": "{\\"expo\\":{\\"name\\":\\"test1\\",\\"sdkVersion\\":\\"39.0.0\\"}}",
          "type": "CODE",
        },
      },
      "dependencies": Object {
        "expo-asset": Object {
          "version": "latest",
        },
      },
      "isDraft": false,
      "manifest": Object {
        "dependencies": Object {
          "expo-asset": "latest",
        },
        "description": "test1 @ Jan 1, 2020",
        "name": "test1",
        "sdkVersion": "39.0.0",
      },
    },
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`git imports repository created with main branch 1`] = `
Array [
  "noop/--/api/v2/snack/save",
  Object {
    "body": Object {
      "code": Object {
        "App.js": Object {
          "contents": "console.log('Hello World');",
          "type": "CODE",
        },
        "app.json": Object {
          "contents": "{\\"expo\\":{\\"name\\":\\"test1\\",\\"sdkVersion\\":\\"39.0.0\\"}}",
          "type": "CODE",
        },
      },
      "dependencies": Object {
        "expo-asset": Object {
          "version": "latest",
        },
      },
      "isDraft": false,
      "manifest": Object {
        "dependencies": Object {
          "expo-asset": "latest",
        },
        "description": "test1 @ Jan 1, 2020",
        "name": "test1",
        "sdkVersion": "39.0.0",
      },
    },
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`git imports repository with app.config.js 1`] = `
Array [
  "noop/--/api/v2/snack/save",
  Object {
    "body": Object {
      "code": Object {
        "App.js": Object {
          "contents": "console.log('Hello World');",
          "type": "CODE",
        },
        "app.config.js": Object {
          "contents": "export default {\\"name\\":\\"some-example\\",\\"sdkVersion\\":\\"39.0.0\\"};",
          "type": "CODE",
        },
      },
      "dependencies": Object {
        "expo-asset": Object {
          "version": "latest",
        },
      },
      "isDraft": false,
      "manifest": Object {
        "dependencies": Object {
          "expo-asset": "latest",
        },
        "description": "some-example @ Jan 1, 2020",
        "name": "some-example",
        "sdkVersion": "39.0.0",
      },
    },
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`git imports repository with assets 1`] = `"noop/--/api/v2/snack/uploadAsset"`;

exports[`git imports repository with assets 2`] = `
Array [
  "noop/--/api/v2/snack/save",
  Object {
    "body": Object {
      "code": Object {
        "App.js": Object {
          "contents": "console.log('Hello World');",
          "type": "CODE",
        },
        "app.json": Object {
          "contents": "{\\"expo\\":{\\"name\\":\\"some-example\\",\\"sdkVersion\\":\\"39.0.0\\"}}",
          "type": "CODE",
        },
        "assets/image.png": Object {
          "contents": "https://myuploadedasset.com/777",
          "type": "ASSET",
        },
      },
      "dependencies": Object {
        "expo-asset": Object {
          "version": "latest",
        },
      },
      "isDraft": false,
      "manifest": Object {
        "dependencies": Object {
          "expo-asset": "latest",
        },
        "description": "some-example @ Jan 1, 2020",
        "name": "some-example",
        "sdkVersion": "39.0.0",
      },
    },
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`git imports repository with custom branch 1`] = `
Array [
  "noop/--/api/v2/snack/save",
  Object {
    "body": Object {
      "code": Object {
        "App.js": Object {
          "contents": "console.log('Hello World');",
          "type": "CODE",
        },
        "app.json": Object {
          "contents": "{\\"expo\\":{\\"name\\":\\"test1\\",\\"sdkVersion\\":\\"39.0.0\\"}}",
          "type": "CODE",
        },
      },
      "dependencies": Object {
        "expo-asset": Object {
          "version": "latest",
        },
      },
      "isDraft": false,
      "manifest": Object {
        "dependencies": Object {
          "expo-asset": "latest",
        },
        "description": "test1@feature-a @ Jan 1, 2020",
        "name": "test1@feature-a",
        "sdkVersion": "39.0.0",
      },
    },
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`git imports repository with no sdkVersion 1`] = `
Array [
  "noop/--/api/v2/snack/save",
  Object {
    "body": Object {
      "code": Object {
        "App.js": Object {
          "contents": "console.log('Hello World');",
          "type": "CODE",
        },
        "app.json": Object {
          "contents": "{\\"expo\\":{\\"name\\":\\"test2\\"}}",
          "type": "CODE",
        },
      },
      "dependencies": Object {
        "expo-asset": Object {
          "version": "latest",
        },
      },
      "isDraft": false,
      "manifest": Object {
        "dependencies": Object {
          "expo-asset": "latest",
        },
        "description": "test2 @ Jan 1, 2020",
        "name": "test2",
        "sdkVersion": "40.0.0",
      },
    },
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`git imports repository with too old sdkVersion 1`] = `
Array [
  "noop/--/api/v2/snack/save",
  Object {
    "body": Object {
      "code": Object {
        "App.js": Object {
          "contents": "console.log('Hello World');",
          "type": "CODE",
        },
        "app.json": Object {
          "contents": "{\\"expo\\":{\\"name\\":\\"test3\\",\\"sdkVersion\\":\\"22.0.0\\"}}",
          "type": "CODE",
        },
      },
      "dependencies": Object {
        "expo-asset": Object {
          "version": "latest",
        },
      },
      "isDraft": false,
      "manifest": Object {
        "dependencies": Object {
          "expo-asset": "latest",
        },
        "description": "test3 @ Jan 1, 2020",
        "name": "test3",
        "sdkVersion": "40.0.0",
      },
    },
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`git imports repository without app.json 1`] = `
Array [
  "noop/--/api/v2/snack/save",
  Object {
    "body": Object {
      "code": Object {
        "App.js": Object {
          "contents": "console.log('Hello World');",
          "type": "CODE",
        },
      },
      "dependencies": Object {
        "expo-asset": Object {
          "version": "latest",
        },
      },
      "isDraft": false,
      "manifest": Object {
        "dependencies": Object {
          "expo-asset": "latest",
        },
        "description": "some-example @ Jan 1, 2020",
        "name": "some-example",
        "sdkVersion": "40.0.0",
      },
    },
    "headers": Object {
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;
